ampd/src/tofnd/grpc.rs

```diff
@@ -0,0 +1,259 @@
+    ) -> Result<Signature>;
+}
+
+pub struct MultisigClient {
+    client: multisig_client::MultisigClient<Channel>,
+    party_uid: String,
+}
+
+impl MultisigClient {
+    pub async fn connect(party_uid: String, url: Url) -> Result<Self> {
+        Ok(Self {
+            party_uid,
+            client: multisig_client::MultisigClient::connect(url.to_string())
+                .await
+                .into_report()
+                .change_context(Error::Grpc)?,
+        })
+    }
+}
+
+#[async_trait]
+impl EcdsaClient for MultisigClient {
+    async fn keygen(&mut self, key_uid: &str) -> Result<PublicKey> {
+        let request = KeygenRequest {
+            key_uid: key_uid.to_string(),
+            party_uid: self.party_uid.to_string(),
+        };
+
+        self.client
+            .keygen(request)
+            .await
+            .and_then(|response| {
+                response
+                    .into_inner()
+                    .keygen_response
+                    .ok_or_else(|| Status::internal("keygen response is empty"))
+            })
+            .into_report()
+            .change_context(Error::Grpc)
+            .and_then(|response| match response {
+                KeygenResponse::PubKey(pub_key) => {
+                    VerifyingKey::from_sec1_bytes(pub_key.as_slice())
+                        .into_report()
+                        .change_context(Error::ParsingFailed)
+                        .attach_printable(format!("{{ invalid_value = {:?} }}", pub_key))
+                        .map(Into::into)
+                }
+                KeygenResponse::Error(error_msg) => Err(TofndError::ExecutionFailed(error_msg))
+                    .into_report()
+                    .change_context(Error::KeygenFailed),
+            })
+    }
+
+    async fn sign(
+        &mut self,
+        key_uid: &str,
+        data: MessageDigest,
+        pub_key: &PublicKey,
+    ) -> Result<Signature> {
+        let request = SignRequest {
+            key_uid: key_uid.to_string(),
+            msg_to_sign: data.to_bytes(),
+            party_uid: self.party_uid.to_string(),
+            pub_key: pub_key.to_bytes(),
+        };
+
+        self.client
+            .sign(request)
+            .await
+            .and_then(|response| {
+                response
+                    .into_inner()
+                    .sign_response
+                    .ok_or_else(|| Status::internal("sign response is empty"))
+            })
+            .into_report()
+            .change_context(Error::Grpc)
+            .and_then(|response| match response {
+                SignResponse::Signature(signature) => Ok(signature),
+                SignResponse::Error(error_msg) => Err(TofndError::ExecutionFailed(error_msg))
+                    .into_report()
+                    .change_context(Error::SignFailed),
+            })
+    }
+}
+
+pub struct SharableEcdsaClient(Arc<Mutex<dyn EcdsaClient>>);
+
+#[allow(dead_code)]
+impl SharableEcdsaClient {
```

Comment:
At this point, can't this also implement `EcdsaClient`? Also, I'd prefer seeing `#[allow(dead_code)]` on the actual methods instead of the struct, cuz the struct is indeed used.

Reply:
unfortunately can't, EcdsaClient takes `&mut self` instead of `&mut self`